src/optimize.rs

```diff
@@ -36,6 +36,15 @@ impl Segment {
 
         mode_bits_count + length_bits_count + data_bits_count
     }
+
+    /// Panics if `&self` is not a valid segment.
+    #[cfg(test)]
+    fn assert_invariants(&self) {
+        // TODO: Consider replacing `begin` and `end` fields with a single
+        // `range: std::ops::Range<usize>` field that would make it impossible
+        // to violate this invariant (well, except for `Range::is_empty`...).
+        assert!(self.begin < self.end);
+    }
```

Comment:
Ideally having `Segment` type preserving the invariant by creating it only via a `new` method returning `Result<Segment,..>` and also `Segments` containing a guaranteed continuous `Vec<Segment>` could be a nice future improvement

Reply:
I agree.

FWIW I've realized that my initial TODO comment was a bit misguided - `Range` can still have `end < begin` except that it defines this to be an empty range.  I've tweaked the comment to instead 1) capture your suggestion and 2) consider replacing `end: usize` with `len: NonZeroUsize`.